text/0000-io-safety.md

```diff
@@ -0,0 +1,429 @@
+it's impossible to bound the set of things that a piece of code might do.
+
+I/O safety is also in this category, for two reasons.
+
+ - I/O safety errors can lead to memory safety errors in the presence of
+   safe wrappers around `mmap` (on platforms with OS-specific APIs allowing
+   them to otherwise be safe).
+
+ - I/O safety errors can also mean that a piece of code can read, write, or
+   delete data used by other parts of the program, without naming them or
+   being given a reference to them. It becomes impossible to bound the set
+   of things a crate can do without knowing the implementation details of all
+   other crates linked into the program.
+
+Raw handles are much like raw pointers into a separate address space; they can
+dangle or be computed in bogus ways. I/O safety is similar to memory safety;
+both prevent spooky-action-at-a-distance, and in both, ownership is the main
+foundation for robust abstractions, so it's natural to use similar safety
+concepts.
+
+[`std::mem::forget` being safe]: https://doc.rust-lang.org/std/mem/fn.forget.html
+[changed to safe]: https://rust-lang.github.io/rfcs/1066-safe-mem-forget.html
+
+## I/O Handles as plain data
+
+The main alternative would be to say that raw handles are plain data, with no
+concept of I/O safety and no inherent relationship to OS resource lifetimes. On
+Unix-like platforms at least, this wouldn't ever lead to memory unsafety or
+undefined behavior.
+
+However, most Rust code doesn't interact with raw handles directly. This is a
+good thing, independently of this RFC, because resources ultimately do have
+lifetimes, so most Rust code will always be better off using higher-level types
+which manage these lifetimes automatically and which provide better ergonomics
+in many other respects. As such, the plain-data approach would at best make raw
+handles marginally more ergonomic for relatively uncommon use cases. This would
+be a small benefit, and may even be a downside, if it ends up encouraging people
+to write code that works with raw handles when they don't need to.
+
+The plain-data approach also wouldn't need any code changes in any crates. The
+I/O safety approach will require changes to Rust code in crates such as
+[`socket2`], [`nix`], and [`mio`] which have APIs involving [`AsRawFd`] and
+[`RawFd`], though the changes can be made gradually across the ecosystem rather
+than all at once.
+
+## New types for `RawFd`/`RawHandle`/`RawSocket`
+
+Some comments on [rust-lang/rust#76969] suggest introducing new wrappers
+around the raw handles. Completely closing the safety loophole would also
+require designing new traits, since `AsRaw*` doesn't have a way to limit the
+lifetime of its return value. This RFC doesn't rule this out, but it would be a
+bigger change.
+
+## I/O safety but not `IoSafe`
+
+The I/O safety concept doesn't depend on `IoSafe` being in `std`. Crates could
+continue to use [`unsafe_io::OwnsRaw`], though that does involve adding a
```

Comment:
I feel like the trait proposed by this RFC would be better conveyed by a name like `OwnsRawFd` rather than `IoSafe`. As noted elsewhere in the RFC, the trait does not fully encapsulate the concept of I/O safety; it's *just* about owning RawFd values.

Reply:
`Fd` is Unix/etc.-specific, which is why earlier versions of this trait are named this just `OwnsRaw`.

I renamed it from `OwnsRaw` to `IoSafe` in response to feedback during the pre-RFC :smile:. I'm happy to rename it to whatever gains consensus here though.

Reply:
> `Fd` is Unix/etc.-specific, which is why earlier versions of this trait are named this just `OwnsRaw`.

Fair; I was hoping for a name that captures the idea of "file". `OwnsRawFile` would work if we call it a `RawFile`, for instance, which might be a good name for an "owned file descriptor or handle".

Reply:
How about just naming it `OwnsOsHandle`, according to Wikipedia, a file descriptor is a kind of handle:
https://en.wikipedia.org/wiki/File_descriptor
> In Unix and Unix-like computer operating systems, a file descriptor (FD, less frequently fildes) is a unique identifier (handle)

Reply:
@programmerjake Given that one major OS uses "handle" as its name for the concept, I think using "handle" as our generic term for the concept just reverses the terminology issue, and has the same problem as using "file descriptor" as our generic name for the concept.

Reply:
`OwnsOsObject`?

Reply:
Rust already uses `Fd` in its terminology. Correcting this name, without also getting rid of `AsRawRd` would introduce inconsistency within Rust nomenclature, and make users wonder what's the difference between an Fd and a Handle, when there's none.

Reply:
Rust doesn't use `Fd` on Windows though.

Currently, `IoSafe` is platform-independent. We could split it into platform-specific variants, similar Rust's `RawFd`/`RawHandle`/`RawSocket` types, but that would increase the boilerplate needed to use it in common cases.

`OwnsOsObject` would work for me. Thoughts?

Reply:
The difficulty in naming this trait is itself a sign of trouble. With the switch away from `IoSafe` and toward `OwnedFd` and `BorrowedFd`, it's no longer necessary to solve this problem :-).